### PR TITLE
add OCI error format support

### PR DESCRIPTION
**Pull request by wy65701436**
1, Leverage go v1.13 new error feature
2, Define genernal error OCI format, so that /v2 API could return a OCI compatible error

Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/pkg/error/errors.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	return "{}"
+}
+
+// Len returns the current number of errors.
+func (errs Errors) Len() int {
+	return len(errs)
+}
+
+// Es ...
+func Es(err error) Errors {
```
Comment (heww) [inline src/pkg/error/errors.go]: `NewErrors` or `FormatError` will be better than `Es`.

File: src/pkg/error/errors.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+	// ErrorPrecondition ...
+	ErrorPrecondition = errors.New("precondition failed")
+
+	// ErrUnknown ...
+	ErrUnknown = errors.New("unknown")
+)
+
+// Code ...
+type Code string
```
Comment (heww) [inline src/pkg/error/errors.go]: `Int` is better than `string` for `Code`

File: src/pkg/error/errors.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	ForbiddenErrorCode
+	// PreconditionErrorCode ...
+	PreconditionErrorCode
+	// GeneralErrorCode ...
+	GeneralErrorCode
+)
+
+// NotFoundError is error for the case of object not found
+func NotFoundError(err error) error {
+	return E(ErrNotFound, Code(strconv.Itoa(ObjectNotFoundErrorCode)), M(err.Error()), D("not found"))
```
Comment (heww) [inline src/pkg/error/errors.go]: When `Code` is int type,  type conversion not needed here.

File: src/pkg/error/errors.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	ErrorPrecondition = errors.New("precondition failed")
+
+	// ErrUnknown ...
+	ErrUnknown = errors.New("unknown")
+)
+
+// Code ...
+type Code string
+
+// M ...
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Please add comment to explain what it is, the same for other definitions

File: src/pkg/error/errors.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		Message: msg,
+		Detail:  e.Detail,
+	}
+}
+
+// Unwrap ...
+func (e Error) Unwrap() error { return e.Err }
+
+// E ...
+func E(args ...interface{}) error {
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Why we need this? What is it used for?

File: src/pkg/error/errors.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+)
+
+var (
+	// ErrNotFound ...
+	ErrNotFound = errors.New("not found")
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Why we need these error definitions? I think the error code is enough to identify the error

File: src/pkg/error/errors.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+type M string
+
+// D ...
+type D string
+
+// Error ...
+type Error struct {
+	Err     error `json:"-"`
+	Code    Code  `json:"code"`
+	Message M     `json:"message"`
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Why don't we use `string` as the type of `Message` and `Detail`?

File: src/pkg/error/errors.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	return &Error{
+		Err:     err,
+		Code:    ObjectConflictErrorCode,
+		Message: err.Error(),
+		Detail:  "conflict",
+	}
+}
+
+// UnauthorizedError is error for the case of unauthorized accessing
+func UnauthorizedError(err error) error {
```
Comment (reasonerjt) [inline src/pkg/error/errors.go]: if the input parm is an instance of `Error` struct, would it be better we `unwrap` the error first?

File: src/common/api/base.go [hunk 9]
```diff
@@ -248,3 +249,47 @@ func (b *BaseAPI) SendPreconditionFailedError(err error) {
...
+//		},
+//		...
+//	]
+// }
+func (b *BaseAPI) SendError(err error) {
+	var statusCode int
+	var send error
+
+	var e pkg_errors.Error
+	if errors.Is(err, e) {
```
Comment (reasonerjt) [inline src/common/api/base.go]: looks like we should use `errors.As` here.

Comment (heww) [reply src/common/api/base.go]: And I have doubts that in controller  how can I check the `error` is a `not found` error returned by dao.

Use `errors.As` and check the code of the error? This is very tricky.

File: src/pkg/error/errors.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+	GeneralErrorCode
+)
+
+// NotFoundError is error for the case of object not found
+func NotFoundError(err error) error {
+	return Error{
+		Err:     err,
+		Code:    ObjectNotFoundErrorCode,
+		Message: err.Error(),
+		Detail:  "not found",
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: The `Message` is a summary message for the error and the `Detail` contains the detail information

Comment (ywk253100) [inline src/pkg/error/errors.go]: The `Message` should be const string for the specific code.

Comment (ywk253100) [reply src/pkg/error/errors.go]: And I should be able to customize the detail information for the specific pre-defined error

File: src/pkg/error/errors.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+// Error returns a human readable error.
+func (e Error) Error() string {
+	if data, err := json.Marshal(e); err == nil {
+		return string(data)
+	}
+	return "{}"
+}
+
+// WithMessage ...
+func (e Error) WithMessage(msg string) Error {
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Seems this function is useless

File: src/pkg/error/errors.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	Err     error  `json:"-"`
+	Code    Code   `json:"code"`
+	Message string `json:"message"`
+	Detail  string `json:"detail,omitempty"`
+}
+
+// Error returns a human readable error.
+func (e Error) Error() string {
+	if data, err := json.Marshal(e); err == nil {
+		return string(data)
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Log the error if marshal failed

File: src/pkg/error/errors.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+	for _, daErr := range errs {
+		var err error
+		switch daErr.(type) {
+		case Error:
+			err = daErr.(Error)
+		default:
+			err = UnknownError(daErr).(Error).WithMessage(err.Error())
+		}
+		tmpErrs.Errors = append(tmpErrs.Errors, Error{
```
Comment (ywk253100) [inline src/pkg/error/errors.go]: Just append the `err` is fine

File: src/pkg/error/errors.go (hunk 10 above)
Comment (wy65701436) [reply src/pkg/error/errors.go]: docker client redenders the message instead of detail, that's why give the error to message.

Comment (wy65701436) [reply src/pkg/error/errors.go]: why message should contain code, use can get it from the code field. And the `withMessage` fun c can let user to customize messages, is it what you want?

File: src/common/api/base.go [hunk 14]
```diff
@@ -248,3 +249,48 @@ func (b *BaseAPI) SendPreconditionFailedError(err error) {
...
+
+	} else {
+		statusCode = http.StatusInternalServerError
+		send = internal_errors.UnknownError(err)
+	}
+
+	b.RenderError(statusCode, internal_errors.NewErrs(send).Error())
+}
+
+func (b *BaseAPI) getStatusCode(code string) int {
```
Comment (renmaosheng) [inline src/common/api/base.go]: capital Get?

File: src/internal/error/errors.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	// NotFoundCode is code for the error of no object found
+	NotFoundCode = "NOT_FOUND"
+	// ConflictCode ...
+	ConflictCode = "CONFLICT"
+	// UnAuthorizedCode ...
+	UnAuthorizedCode = "UNAUTHORIZED"
+	// BadRequestCode ...
+	BadRequestCode = "BAD_REQUEST"
+	// ForbiddenCode ...
+	ForbiddenCode = "FORBIDDER"
```
Comment (renmaosheng) [inline src/internal/error/errors.go]: FORBIDDEN?

File: src/internal/error/errors.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// ForbiddenError is error for the case of forbidden
+func ForbiddenError(err error) *Error {
+	return New(err).WithCode(ForbiddenCode).WithMessage("forbidden")
+}
+
+// PreconditionFailedError is error for the case of precondition failed
+func PreconditionFailedError(err error) *Error {
+	return New(err).WithCode(PreconditionCode).WithMessage("preconfition")
```
Comment (renmaosheng) [inline src/internal/error/errors.go]: precondition?

File: src/common/api/base.go (hunk 14 above)
Comment (ywk253100) [reply src/common/api/base.go]: It's a internal method, so the capital Get isn't needed

File: src/internal/error/errors.go (hunk 15 above)
Comment (ywk253100) [reply src/internal/error/errors.go]: Fix included in https://github.com/goharbor/harbor/pull/10366

File: src/internal/error/errors.go (hunk 16 above)
Comment (ywk253100) [reply src/internal/error/errors.go]: Fix included in https://github.com/goharbor/harbor/pull/10366

**Pull request merged**